docs/RFCS/20190318_error_handling.md

```diff
@@ -0,0 +1,2090 @@
+| (CRDB) proposed new error objects              | good      | good        | good                   | good    | good         | good                 |
+
+This failure by the current code to meet all our requirements is the main motivation for this work.
+
+# Guide-level explanation
+
+The package is `github.com/cockroachdb/cockroach/pkg/errors`.
+
+Table of contents:
+
+- [Instantiating new errors](#Instantiating-new-errors)
+- [Decorating existing errors](#Decorating-existing-errors)
+- [Utility features](#Utility-features)
+  - [Adding `context`](#Adding-context)
+  - [Formatting variants](#Formatting-variants)
+  - [Safe details](#Safe-details)
+  - [Additional PostgreSQL error fields](#Additional-PostgreSQL-error-fields)
+  - [Telemetry keys](#Telemetry-keys)
+  - [Depth variants](#Depth-variants)
+- [Handling chains of error causes](#Handling-chains-of-error-causes)
+  - [Accessing the cause](#Accessing-the-cause)
+  - [Preservation of causes across the wire](#Preservation-of-causes-across-the-wire)
+  - [Barrier errors](#Barrier-errors)
+  - [Multiple causes](#Multiple-causes)
+  - [Identification of causes](#Identification-of-causes)
+  - [Error equivalence and markers](#Error-equivalence-and-markers)
+- [Promotion and preservation during formatting](#Promotion-and-preservation-during-formatting)
+- [What comes out of an error?](#What-comes-out-of-an-error)
+
+## Instantiating new errors
+
+Instantiating a new error can be as simple as `errors.New("hello")` or
+`errors.Errorf("hello %s", "world")`.
+
+To equip a code useful to PostgreSQL clients when constructing a new
+error, the following form is preferred:
+`errors.New("hello").WithDefaultCode(pgerror.CodeSyntaxError)`
+
+The library is compatible with existing protobuf error objects, so
+instantiating, for example, with `err :=
+&roachpb.RangeRetryError{Reason: "hello"}` is also valid.
+
+## Decorating existing errors
+
+Adding some words of context can be as simple as `errors.Wrap(err,
+"hello")`, although the form `Wrapx()` is preferred (see next section).
+
+To add a code useful to PostgreSQL clients, one can use
+e.g. `errors.WithDefaultCode(err, pgerror.CodeSyntaxError)`.
+
+The new code is only used if the original error did not provide a code
+already. This is suitable under the assumption that the code closest
+to where the error originated is the most able to provide a precise
+error code. (This assumption was manually evaluated to hold in the
+code audited so far.) A further iteration of the error handling
+library may provide richer opt-in code override semantics.
+
+## Utility features
+
+The following features are opt-in and can be used to enhance the
+quality of error details included in telemetry or available for
+troubleshooting.
+
+### Adding `context`
+
+When wrapping an error with `errors.WithCtx(err, ctx)`, the logging
```

Comment:
This scares me - context lifetimes are already complicated enough - we really don't want them to escape into errors if we can avoid it.